macros/PGauxiliaryFunctions.pl

```diff
@@ -127,42 +127,162 @@ sub lcm {
+# Arguments should be array references to arrays of integers.
+# Returns an n-tuple of relatively prime integers,
+# each one coming from the corresponding array.
+# Random selection is uniform among all possible tuples that are relatively prime.
+# Does not consider (0,0) to be relatively prime.
+# In array context, returns an array. Otherwise, an array ref.
+# Use like:
+# random_coprime([1..9],[1..9]) to output maybe (2,9) or (1,1) but not (6,8)
+# random_coprime([-9..-1,1..9],[1..9],[1..9]) to output maybe (-3,7,4), (-1,1,1), or (-2,2,3) but not (-2,2,4)
+# random_pairwise_coprime([-9..-1,1..9],[1..9],[1..9]) to output maybe (-3,7,4) or (-1,1,1) but not (-2,2,3)
+# WARNING: random_coprime() will use a lot of memory and CPU resources if used with too many/too large arguments.
+# For example, random_coprime([-20..20],[-20..20],[-20..20],[-20..20],[-20..20]) involves processing 41^5 arrays.
+# Consider using random_pairwise_coprime() instead. Or breaking things up like:
+# random_coprime([-20..20],[-20..20]),random_coprime([-20..20],[-20..20],[-20..20])
+
+
+# ^ function random_coprime
+# ^uses gcd
+sub random_coprime {
+	# Expect first argument to be an array reference
+	my $c = shift;
+	my @candidates = @$c if $c;
+	# @candidates has numbers on the first iteration
+	# On subsequent iterations it has two array references
+	# The first of these is full of array references to tuples where we already know the gcf is 1
+	# The second is full of array references to tuples where the gcf is not 1, but these may become usable on later iterations
+	# If it has numbers, initialize the two array refs
+	if (ref $candidates[0] eq '') {
+		do {warn "Unable to find a coprime tuple from input"; return;} unless (@candidates);
+		my @refcandidates = ([],[]);
+		for my $i (@candidates) {
+			if (abs($i) == 1) {
+				push @{$refcandidates[0]},[$i];
+			} else {
+				push @{$refcandidates[1]},[$i];
+			}
+		}
+		return random_coprime([@refcandidates],@_);
+	} elsif (ref $candidates[0] eq 'ARRAY') {
+		# Expect second argument to be an array reference to an array of integers, if present
+		my $n = shift;
+		my @newcomers = @$n if ($n);
+		if (@newcomers) {
+			# Cross @candidates with @newcomers to make @newcandidates
+			my @newcandidates = ([],[]);
+			for my $i (@{$candidates[0]}) {
+				for my $j (@newcomers) {
+					push @{$newcandidates[0]}, [@{$i}, $j];
+				}
+			}
+			for my $i (@{$candidates[1]}) {
+				for my $j (@newcomers) {
+					# next three lines are to avoid asking for gcf of all-zero set
+					my $hasnonzero = 0;
+					for my $k (@{$i}) {do {$hasnonzero = 1; last;} if ($k != 0)};
+					do {push @{$newcandidates[1]}, [@{$i}, $j]; next} unless ($hasnonzero or $j != 0);
+					if (gcf($j,@{$i}) == 1) {
+						push @{$newcandidates[0]}, [@{$i}, $j];
+					} else {
+						push @{$newcandidates[1]}, [@{$i}, $j];
+					}
+				}
+			}
+			do {warn "Unable to find a coprime tuple from input"; return;} unless (@{$newcandidates[0]},@{$newcandidates[1]});
```

Comment:
I have noticed that this causes a message `Useless use of a variable in void context` in the apache2 error log.  I am not sure what you intended here, but the `unless` conditional doesn't make sense and is what is causing the message.  The second element of the list (`@{$newcandidates[1]}`) is what is in the void context and is being ignored in the conditional.  Did you mean `unless (@{$newcandidates[0]} || @{$newcandidates[1]})`?